### ISSUE TITLE
Issue with ServiceInstance UpdateRequests field

### ISSUE THREAD

**Issue by ash2k** [kubernetes-incubator/service-catalog#1454]
1. It feels like this field should be optional and hence a pointer. Being non-pointer implies the user has to provide it. But if omitted from YAML/JSON create operation works fine. Does that mean it is actually optional? It contradicts this point from [api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required):
> The API server should not allow POSTing or PUTing a resource with this field unset.
2. It causes inconvenience - if code or a user wants to send an update to overwrite the current state of the object unconditionally, it has to know the current value of the field.
3. I wonder why this field is needed? There is no code that uses it. That implies (having read the comment on the field) that the only purpose of the field is to generate an update event to make the controller reprocess the object. The same can be achieved by sending an update with exactly the same object - it will generate an event. Even if it was not true, there is always an option of adding an annotation with random name.

Maybe someone from api machinery can review the types before beta is released?

**Comment by nilebox** [comment]
> It feels like this field should be optional and hence a pointer

Thanks for the link to k8s API docs, agreed, we should make `updateRequests` field optional with a pointer.

> if code or a user wants to send an update to overwrite the current state of the object unconditionally, it has to know the current value of the field

This is true for any fields including the ones that user hasn't touched, isn't it? You should always pass the entire object in PUT request, it's normal. There is also a PATCH request, where can omit unchanged fields IIRC.

> the only purpose of the field is to generate an update event to make the controller reprocess the object

Correct.

> The same can be achieved by sending an update with exactly the same object - it will generate an event

We need to update the `Generation` field to reliably reconcile the updated resource. In Service Catalog it is being updated manually only when the spec changes, see https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/registry/servicecatalog/instance/strategy.go#L161-L168
So neither empty update, nor annotation update will lead to updating a `Generation`.

> there is always an option of adding an annotation with random name.

The goal of `updateRequests` is to make enforced reconciliation loops user-friendly. Asking users to manually update metadata with random values is not user-friendly. We can solve it by adding a `/forceUpdate` subresource, but it still won't be supported by `kubectl` I guess? We can potentially develop a `kubectl` plugin for this, but the overall solution seems pretty complicated.

The better design (both from the UX and system design perspective) is definitely watching on Secret updates, but they are not properly supported by Kubernetes yet.

**Comment by nilebox** [comment]
**TL;DR** `updateRequests` field was added to let the user force OSB update request when the parameters in a referenced secret object were updated. When we discussed this design, the conclusion was that this a good enough simple temporary solution until Kubernetes will add support for Secret "bulk watch" (see https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/bulk_watch.md and https://github.com/kubernetes/ingress-nginx/issues/816#issuecomment-309630254), which is still not implemented AFAIK. The long-term design is to support Secret watch instead of manual `updateRequests` or any other hack, I believe.

@pmorie is that correct?

**Comment by pmorie** [comment]
> updateRequests field was added to let the user force OSB update request when the parameters in a referenced secret object were updated

That's correct; you need something to bump in the event that you _just_ change the out-of-band parameters.  I don't think we should encourage updates to the same state as a mechanism for doing this.

**Comment by pmorie** [comment]
Why don't we make this optional, and if the value received is zero, just carry the value forward from the last state in the API server?

I don't think this needs to be a pointer personally.

**Comment by nilebox** [comment]
@pmorie agree. by looking at the k8s core I haven't found consistent usage of pointers for the fields marjed with `+optional`. Also, switching to pointer in the future doesn't break the REST API, so we can make this change anytime if needed.
